src/xmipp/libraries/reconstruction_cuda/cuda_angular_sph_alignment.cpp

```diff
@@ -0,0 +1,648 @@
+#include "reconstruction_adapt_cuda/angular_sph_alignment_gpu.h"
+#include "cuda_angular_sph_alignment.h"
+#include "cuda_angular_sph_alignment.cu"
+#include "cuda_volume_deform_sph_defines.h"//TODO
+// Standard includes
+#include <iterator>
+#include <stdexcept>
+#include <stdio.h>
+#include <iostream>
+#include <exception>
+// Thrust includes
+#include <thrust/reduce.h>
+#include <thrust/device_vector.h>
+
+
+namespace AngularAlignmentGpu {
+
+// Common functions
+template<typename T>
+cudaError cudaMallocAndCopy(T** target, const T* source, size_t numberOfElements, size_t memSize = 0)
+{
+    size_t elemSize = numberOfElements * sizeof(T);
+    memSize = memSize == 0 ? elemSize : memSize * sizeof(T);
+
+    cudaError err = cudaSuccess;
+    if ((err = cudaMalloc(target, memSize)) != cudaSuccess) {
+        *target = NULL;
+        return err;
+    }
+
+    if ((err = cudaMemcpy(*target, source, elemSize, cudaMemcpyHostToDevice)) != cudaSuccess) {
+        cudaFree(*target);
+        *target = NULL;
+    }
+
+    if (memSize > elemSize) {
+        cudaMemset((*target) + numberOfElements, 0, memSize - elemSize);
+    }
+
+    return err;
+}
+
+#define processCudaError() (_processCudaError(__FILE__, __LINE__))
+void _processCudaError(const char* file, int line) 
+{
+    cudaError_t err = cudaGetLastError();
+    if (err != cudaSuccess) {
+        fprintf(stderr, "File: %s: line %d\nCuda error: %s\n", file, line, cudaGetErrorString(err));
+        exit(err);
+    }
+}
+
+// Copies data from CPU to the GPU and at the same time transforms from
+// type 'U' to type 'T'. Works only for numeric types
+template<typename Target, typename Source>
+void transformData(Target** dest, Source* source, size_t n, bool mallocMem = true)
+{
+    std::vector<Target> tmp(source, source + n);
+
+    if (mallocMem){
+        if (cudaMalloc(dest, sizeof(Target) * n) != cudaSuccess){
+            processCudaError();
+        }
+    }
+
+    if (cudaMemcpy(*dest, tmp.data(), sizeof(Target) * n, cudaMemcpyHostToDevice) != cudaSuccess){
+        processCudaError();
+    }
+}
+
+// AngularSphAlignment methods
+
+AngularSphAlignment::AngularSphAlignment()
+{
+}
+
+AngularSphAlignment::~AngularSphAlignment() 
+{
+    if (program->useFakeKernel) {
+    } else {
+        cudaFree(dVolData);
+        cudaFree(dRotation);
+        cudaFree(dZshParams);
+        cudaFree(dClnm);
+        cudaFree(dVolMask);
+        cudaFree(dProjectionPlane);
+    }
+}
+
+void AngularSphAlignment::associateWith(ProgAngularSphAlignmentGpu* prog) 
+{
+    program = prog;
+}
+
+static dim3 grid;
+static dim3 block;
+
+void AngularSphAlignment::setupConstantParameters() 
+{
+    if (program == nullptr)
+        throw new std::runtime_error("AngularSphAlignment not associated with the program!");
+
+    // kernel arguments
+    this->Rmax2 = program->RmaxDef * program->RmaxDef;
+    this->iRmax = 1.0 / program->RmaxDef;
+    setupImageMetaData(program->V);
+
+    if (program->useFakeKernel) {
```

Comment:
what is the reason behind having cpu computation in the  gpu version of the program?